File: conformance/go/decode.go
// Package main is the Graft conformance port for Go: a zero-dependency decoder
// plus a runner that matches spec/golden/*.bin against the .meta.json sidecars.
//
// This file holds the value graph and the two-pass heap decoder (FORMAT.md §4):
// reference types are created as empty pointers first, then filled, so shared
// identity and cycles are restored rather than copied.
package main

import (
	"encoding/binary"
	"fmt"
	"math"
	"math/big"
)

// --- value graph -----------------------------------------------------------
//
// A decoded Value is held in an interface{}. Reference types are pointers so
// they carry identity (shared refs / cycles); leaves are plain values.

// Undefined models JS `undefined`, kept distinct from nil (JS `null`).
type Undefined struct{}

var undefinedVal = Undefined{}

// BigInt is an arbitrary-precision integer as its canonical decimal string.
type BigInt string

type Bytes []byte    // JS ArrayBuffer
type DataView []byte // JS DataView (viewed window)

type TypedArray struct {
	ElementType string
	Data        []byte
}

type Date struct {
	UnixMS     int64
	SubMsNanos int64
}

type RegExp struct {
	Source string
	Flags  string
}

type URL struct {
	Href string
}

// Symbol is referenced by pointer so unique symbols compare by identity.
type Symbol struct {
	Kind  string // "registered" | "unique" | "well_known"
	Value string
}

// Entry is one Object/Error property: Key is a string or *Symbol.
type Entry struct {
	Key   interface{}
	Value interface{}
}

type Array struct{ Items []interface{} }
type Object struct{ Entries []Entry }

type Pair struct{ Key, Value interface{} }
type MapV struct{ Entries []Pair }
type SetV struct{ Values []interface{} }

type ErrorV struct {
	Name     string
	Message  string
	HasCause bool
	Cause    interface{}
	Extra    []Entry
}

// --- reader ----------------------------------------------------------------

type reader struct {
	data []byte
	pos  int
}

func (r *reader) take(n int) ([]byte, error) {
	if r.pos+n > len(r.data) {
		return nil, fmt.Errorf("EOF")
	}
	out := r.data[r.pos : r.pos+n]
	r.pos += n
	return out, nil
}

func (r *reader) u8() (byte, error) {
	b, err := r.take(1)
	if err != nil {
		return 0, err
	}
	return b[0], nil
}

func (r *reader) uvarint() (uint64, error) {
	var result uint64
	var shift uint
	for {
		b, err := r.u8()
		if err != nil {
			return 0, err
		}
		result |= uint64(b&0x7f) << shift
		if b&0x80 == 0 {
			break
		}
		shift += 7
		if shift >= 64 {
			return 0, fmt.Errorf("uvarint overflows 64 bits")
		}
	}
	return result, nil
}

func (r *reader) uvarintInt() (int, error) {
	v, err := r.uvarint()
	return int(v), err
}

func (r *reader) svarint() (int64, error) {
	z, err := r.uvarint()
	if err != nil {
		return 0, err
	}
	if z&1 == 0 {
		return int64(z >> 1), nil
	}
	return -int64((z + 1) >> 1), nil
}

// uvarintGroups returns a LEB128 value's 7-bit groups (little-endian) for bigints.
func (r *reader) uvarintGroups() ([]byte, error) {
	var groups []byte
	for {
		b, err := r.u8()
		if err != nil {
			return nil, err
		}
		groups = append(groups, b&0x7f)
		if b&0x80 == 0 {
			break
		}
	}
	return groups, nil
}

func (r *reader) f64() (float64, error) {
	b, err := r.take(8)
	if err != nil {
		return 0, err
	}
	return math.Float64frombits(binary.LittleEndian.Uint64(b)), nil
}

func (r *reader) str() (string, error) {
	n, err := r.uvarintInt()
	if err != nil {
		return "", err
	}
	b, err := r.take(n)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// --- node decoding ---------------------------------------------------------

func bigIntDecimal(sign byte, groups []byte) string {
	n := big.NewInt(0)
	base := big.NewInt(128)
	tmp := new(big.Int)
	for i := len(groups) - 1; i >= 0; i-- {
		n.Mul(n, base)
		n.Add(n, tmp.SetInt64(int64(groups[i])))
	}
	s := n.String()
	if sign == 1 && s != "0" {
		s = "-" + s
	}
	return s
}

var elementTypeNames = map[byte]string{
	0: "Uint8", 1: "Uint8Clamped", 2: "Uint16", 3: "Uint32",
	4: "Int8", 5: "Int16", 6: "Int32",
	7: "Float32", 8: "Float64", 9: "BigInt64", 10: "BigUint64",
}

type rawEntry struct {
	isSym  bool
	keyStr string
	keyRef int
	valRef int
}

func readEntries(r *reader) ([]rawEntry, error) {
	n, err := r.uvarintInt()
	if err != nil {
		return nil, err
	}
	entries := make([]rawEntry, 0, n)
	for i := 0; i < n; i++ {
		kind, err := r.u8()
		if err != nil {
			return nil, err
		}
		var e rawEntry
		if kind == 0 {
			if e.keyStr, err = r.str(); err != nil {
				return nil, err
			}
		} else {
			e.isSym = true
			if e.keyRef, err = r.uvarintInt(); err != nil {
				return nil, err
			}
		}
		if e.valRef, err = r.uvarintInt(); err != nil {
			return nil, err
		}
		entries = append(entries, e)
	}
	return entries, nil
}

func readRefs(r *reader) ([]int, error) {
	n, err := r.uvarintInt()
	if err != nil {
		return nil, err
	}
	refs := make([]int, 0, n)
	for i := 0; i < n; i++ {
		ref, err := r.uvarintInt()
		if err != nil {
			return nil, err
		}
		refs = append(refs, ref)
	}
	return refs, nil
}

// filler populates a container in the second pass using the completed heap.
type filler func(heap []interface{})

func resolveKey(e rawEntry, heap []interface{}) interface{} {
	if e.isSym {
		return heap[e.keyRef]
	}
	return e.keyStr
}

func readNode(r *reader) (interface{}, filler, error) {
	tag, err := r.u8()
	if err != nil {
		return nil, nil, err
	}
	switch tag {
	case 0:
		return nil, nil, nil
	case 1:
		return undefinedVal, nil, nil
	case 2:
		return false, nil, nil
	case 3:
		return true, nil, nil
	case 4:
		v, err := r.svarint()
		return v, nil, err
	case 5:
		v, err := r.f64()
		return v, nil, err
	case 6:
		sign, err := r.u8()
		if err != nil {
			return nil, nil, err
		}
		groups, err := r.uvarintGroups()
		if err != nil {
			return nil, nil, err
		}
		return BigInt(bigIntDecimal(sign, groups)), nil, nil
	case 7:
		v, err := r.str()
		return v, nil, err

	case 10, 11, 12:
		kind := map[byte]string{10: "registered", 11: "unique", 12: "well_known"}[tag]
		v, err := r.str()
		if err != nil {
			return nil, nil, err
		}
		return &Symbol{Kind: kind, Value: v}, nil, nil

	case 20: // Array
		refs, err := readRefs(r)
		if err != nil {
			return nil, nil, err
		}
		arr := &Array{}
		return arr, func(heap []interface{}) {
			for _, i := range refs {
				arr.Items = append(arr.Items, heap[i])
			}
		}, nil
	case 21: // Object
		entries, err := readEntries(r)
		if err != nil {
			return nil, nil, err
		}
		obj := &Object{}
		return obj, func(heap []interface{}) {
			for _, e := range entries {
				obj.Entries = append(obj.Entries, Entry{Key: resolveKey(e, heap), Value: heap[e.valRef]})
			}
		}, nil
	case 22, 30: // Map / WeakMap
		n, err := r.uvarintInt()
		if err != nil {
			return nil, nil, err
		}
		pairs := make([][2]int, 0, n)
		for i := 0; i < n; i++ {
			k, err := r.uvarintInt()
			if err != nil {
				return nil, nil, err
			}
			v, err := r.uvarintInt()
			if err != nil {
				return nil, nil, err
			}
			pairs = append(pairs, [2]int{k, v})
		}
		m := &MapV{}
		return m, func(heap []interface{}) {
			for _, p := range pairs {
				m.Entries = append(m.Entries, Pair{Key: heap[p[0]], Value: heap[p[1]]})
			}
		}, nil
	case 23, 31: // Set / WeakSet
		refs, err := readRefs(r)
		if err != nil {
			return nil, nil, err
		}
		s := &SetV{}
		return s, func(heap []interface{}) {
			for _, i := range refs {
				s.Values = append(s.Values, heap[i])
			}
		}, nil

	case 40: // Date
		ms, err := r.svarint()
		if err != nil {
			return nil, nil, err
		}
		sub, err := r.svarint()
		if err != nil {
			return nil, nil, err
		}
		return Date{UnixMS: ms, SubMsNanos: sub}, nil, nil
	case 41: // Bytes
		n, err := r.uvarintInt()
		if err != nil {
			return nil, nil, err
		}
		b, err := r.take(n)
		if err != nil {
			return nil, nil, err
		}
		return Bytes(append([]byte(nil), b...)), nil, nil
	case 42: // TypedArray
		et, err := r.u8()
		if err != nil {
			return nil, nil, err
		}
		n, err := r.uvarintInt()
		if err != nil {
			return nil, nil, err
		}
		b, err := r.take(n)
		if err != nil {
			return nil, nil, err
		}
		name, ok := elementTypeNames[et]
		if !ok {
			return nil, nil, fmt.Errorf("unknown element type: %d", et)
		}
		return TypedArray{ElementType: name, Data: append([]byte(nil), b...)}, nil, nil
	case 43: // RegExp
		src, err := r.str()
		if err != nil {
			return nil, nil, err
		}
		flags, err := r.str()
		if err != nil {
			return nil, nil, err
		}
		return RegExp{Source: src, Flags: flags}, nil, nil
	case 44: // Url
		href, err := r.str()
		if err != nil {
			return nil, nil, err
		}
		return URL{Href: href}, nil, nil
	case 45: // DataView
		n, err := r.uvarintInt()
		if err != nil {
			return nil, nil, err
		}
		b, err := r.take(n)
		if err != nil {
			return nil, nil, err
		}
		return DataView(append([]byte(nil), b...)), nil, nil
	case 46: // Error
		name, err := r.str()
		if err != nil {
			return nil, nil, err
		}
		message, err := r.str()
		if err != nil {
			return nil, nil, err
		}
		flags, err := r.u8()
		if err != nil {
			return nil, nil, err
		}
		hasCause := flags&1 != 0
		causeRef := -1
		if hasCause {
			if causeRef, err = r.uvarintInt(); err != nil {
				return nil, nil, err
			}
		}
		entries, err := readEntries(r)
		if err != nil {
			return nil, nil, err
		}
		e := &ErrorV{Name: name, Message: message, HasCause: hasCause}
		return e, func(heap []interface{}) {
			if hasCause {
				e.Cause = heap[causeRef]
			}
			for _, en := range entries {
				e.Extra = append(e.Extra, Entry{Key: resolveKey(en, heap), Value: heap[en.valRef]})
			}
		}, nil
	}
	return nil, nil, fmt.Errorf("unknown tag: %d", tag)
}

// Decode parses a Graft stream into a native value graph.
func Decode(data []byte) (interface{}, error) {
	r := &reader{data: data}
	magic, err := r.take(4)
	if err != nil {
		return nil, err
	}
	if string(magic) != "GRF1" {
		return nil, fmt.Errorf("bad magic: not a Graft file")
	}
	version, err := r.u8()
	if err != nil {
		return nil, err
	}
	if version != 1 {
		return nil, fmt.Errorf("unsupported version: %d", version)
	}
	root, err := r.uvarintInt()
	if err != nil {
		return nil, err
	}
	count, err := r.uvarintInt()
	if err != nil {
		return nil, err
	}

	heap := make([]interface{}, count)
	var fills []filler
	for i := 0; i < count; i++ {
		v, fill, err := readNode(r)
		if err != nil {
			return nil, err
		}
		heap[i] = v
		if fill != nil {
			fills = append(fills, fill)
		}
	}
	for _, fill := range fills {
		fill(heap)
	}
	if root < 0 || root >= len(heap) {
		return nil, fmt.Errorf("root index %d out of range", root)
	}
	return heap[root], nil
}

File: conformance/rust/src/decode.rs
//! Graft binary decoder. Two-pass over the heap (FORMAT.md §4): build every node
//! (reference types as empty `Rc` placeholders), then fill containers by cloning
//! the now-populated heap slots, so shared identity and cycles are restored.

use crate::value::{ErrorObj, Key, SymbolKind, Value};
use std::cell::RefCell;
use std::rc::Rc;

struct Reader<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Reader<'a> {
    fn new(data: &'a [u8]) -> Self {
        Reader { data, pos: 0 }
    }
    fn take(&mut self, n: usize) -> Result<&'a [u8], String> {
        if self.pos + n > self.data.len() {
            return Err("EOF".into());
        }
        let s = &self.data[self.pos..self.pos + n];
        self.pos += n;
        Ok(s)
    }
    fn u8(&mut self) -> Result<u8, String> {
        Ok(self.take(1)?[0])
    }
    fn uvarint(&mut self) -> Result<u64, String> {
        let mut result = 0u64;
        let mut shift = 0u32;
        loop {
            let byte = self.u8()?;
            result |= ((byte & 0x7f) as u64) << shift;
            if byte & 0x80 == 0 {
                break;
            }
            shift += 7;
            if shift >= 64 {
                return Err("uvarint overflows 64 bits".into());
            }
        }
        Ok(result)
    }
    fn uvarint_usize(&mut self) -> Result<usize, String> {
        Ok(self.uvarint()? as usize)
    }
    fn svarint(&mut self) -> Result<i64, String> {
        let z = self.uvarint()?;
        Ok(if z & 1 == 0 {
            (z >> 1) as i64
        } else {
            -(((z + 1) >> 1) as i64)
        })
    }
    /// Reads a LEB128 value as its 7-bit groups (little-endian) for bigints.
    fn uvarint_groups(&mut self) -> Result<Vec<u8>, String> {
        let mut groups = Vec::new();
        loop {
            let b = self.u8()?;
            groups.push(b & 0x7f);
            if b & 0x80 == 0 {
                break;
            }
        }
        Ok(groups)
    }
    fn f64(&mut self) -> Result<f64, String> {
        let b = self.take(8)?;
        let arr: [u8; 8] = b.try_into().map_err(|_| "short f64".to_string())?;
        Ok(f64::from_le_bytes(arr))
    }
    fn string(&mut self) -> Result<String, String> {
        let n = self.uvarint_usize()?;
        let bytes = self.take(n)?;
        String::from_utf8(bytes.to_vec()).map_err(|_| "invalid utf-8".into())
    }
}

fn groups_to_decimal(sign: u8, groups: &[u8]) -> String {
    // value = sum(groups[i] * 128^i); groups are little-endian 7-bit payloads.
    use num_bigint::BigUint;
    let b128 = BigUint::from(128u32);
    let mut n = BigUint::from(0u32);
    for &g in groups.iter().rev() {
        n = n * &b128 + BigUint::from(g);
    }
    let s = n.to_str_radix(10);
    if sign == 1 && s != "0" {
        format!("-{}", s)
    } else {
        s
    }
}

fn element_type_name(code: u8) -> Result<&'static str, String> {
    Ok(match code {
        0 => "Uint8",
        1 => "Uint8Clamped",
        2 => "Uint16",
        3 => "Uint32",
        4 => "Int8",
        5 => "Int16",
        6 => "Int32",
        7 => "Float32",
        8 => "Float64",
        9 => "BigInt64",
        10 => "BigUint64",
        other => return Err(format!("unknown element type: {}", other)),
    })
}

enum RawKey {
    Str(String),
    Sym(usize),
}
struct RawEntry {
    key: RawKey,
    val: usize,
}

enum Filler {
    Array(Rc<RefCell<Vec<Value>>>, Vec<usize>),
    Object(Rc<RefCell<Vec<(Key, Value)>>>, Vec<RawEntry>),
    Map(Rc<RefCell<Vec<(Value, Value)>>>, Vec<(usize, usize)>),
    Set(Rc<RefCell<Vec<Value>>>, Vec<usize>),
    Error(Rc<RefCell<ErrorObj>>, Option<usize>, Vec<RawEntry>),
}

fn read_entries(r: &mut Reader) -> Result<Vec<RawEntry>, String> {
    let n = r.uvarint_usize()?;
    let mut entries = Vec::with_capacity(n);
    for _ in 0..n {
        let kind = r.u8()?;
        let key = if kind == 0 {
            RawKey::Str(r.string()?)
        } else {
            RawKey::Sym(r.uvarint_usize()?)
        };
        let val = r.uvarint_usize()?;
        entries.push(RawEntry { key, val });
    }
    Ok(entries)
}

fn read_node(r: &mut Reader) -> Result<(Value, Option<Filler>), String> {
    let tag = r.u8()?;
    Ok(match tag {
        0 => (Value::Null, None),
        1 => (Value::Undefined, None),
        2 => (Value::Bool(false), None),
        3 => (Value::Bool(true), None),
        4 => (Value::Int(r.svarint()?), None),
        5 => (Value::Float(r.f64()?), None),
        6 => {
            let sign = r.u8()?;
            let groups = r.uvarint_groups()?;
            (Value::BigInt(groups_to_decimal(sign, &groups)), None)
        }
        7 => (Value::Str(r.string()?), None),

        10 => (sym(r, "registered")?, None),
        11 => (sym(r, "unique")?, None),
        12 => (sym(r, "well_known")?, None),

        20 => {
            let n = r.uvarint_usize()?;
            let mut refs = Vec::with_capacity(n);
            for _ in 0..n {
                refs.push(r.uvarint_usize()?);
            }
            let rc = Rc::new(RefCell::new(Vec::new()));
            (Value::Array(rc.clone()), Some(Filler::Array(rc, refs)))
        }
        21 => {
            let entries = read_entries(r)?;
            let rc = Rc::new(RefCell::new(Vec::new()));
            (Value::Object(rc.clone()), Some(Filler::Object(rc, entries)))
        }
        22 | 30 => {
            let n = r.uvarint_usize()?;
            let mut pairs = Vec::with_capacity(n);
            for _ in 0..n {
                pairs.push((r.uvarint_usize()?, r.uvarint_usize()?));
            }
            let rc = Rc::new(RefCell::new(Vec::new()));
            (Value::Map(rc.clone()), Some(Filler::Map(rc, pairs)))
        }
        23 | 31 => {
            let n = r.uvarint_usize()?;
            let mut refs = Vec::with_capacity(n);
            for _ in 0..n {
                refs.push(r.uvarint_usize()?);
            }
            let rc = Rc::new(RefCell::new(Vec::new()));
            (Value::Set(rc.clone()), Some(Filler::Set(rc, refs)))
        }

        40 => {
            let unix_ms = r.svarint()?;
            let sub_ms_nanos = r.svarint()?;
            (
                Value::Date {
                    unix_ms,
                    sub_ms_nanos,
                },
                None,
            )
        }
        41 => {
            let n = r.uvarint_usize()?;
            (Value::Bytes(r.take(n)?.to_vec()), None)
        }
        42 => {
            let et = r.u8()?;
            let n = r.uvarint_usize()?;
            let raw = r.take(n)?.to_vec();
            (
                Value::TypedArray {
                    element_type: element_type_name(et)?.to_string(),
                    data: raw,
                },
                None,
            )
        }
        43 => (
            Value::Regex {
                source: r.string()?,
                flags: r.string()?,
            },
            None,
        ),
        44 => (Value::Url(r.string()?), None),
        45 => {
            let n = r.uvarint_usize()?;
            (Value::DataView(r.take(n)?.to_vec()), None)
        }
        46 => {
            let name = r.string()?;
            let message = r.string()?;
            let flags = r.u8()?;
            let has_cause = flags & 1 != 0;
            let cause_ref = if has_cause {
                Some(r.uvarint_usize()?)
            } else {
                None
            };
            let entries = read_entries(r)?;
            let rc = Rc::new(RefCell::new(ErrorObj {
                name,
                message,
                has_cause,
                cause: Value::Null,
                extra: Vec::new(),
            }));
            (
                Value::Error(rc.clone()),
                Some(Filler::Error(rc, cause_ref, entries)),
            )
        }

        other => return Err(format!("unknown tag: {}", other)),
    })
}

fn sym(r: &mut Reader, kind: &str) -> Result<Value, String> {
    Ok(Value::Symbol(Rc::new(SymbolKind {
        kind: kind.to_string(),
        value: r.string()?,
    })))
}

fn resolve_key(key: &RawKey, heap: &[Value]) -> Key {
    match key {
        RawKey::Str(s) => Key::Str(s.clone()),
        RawKey::Sym(i) => Key::Sym(heap[*i].clone()),
    }
}

fn fill(filler: Filler, heap: &[Value]) {
    match filler {
        Filler::Array(rc, refs) => {
            let mut v = rc.borrow_mut();
            for i in refs {
                v.push(heap[i].clone());
            }
        }
        Filler::Object(rc, entries) => {
            let mut v = rc.borrow_mut();
            for e in entries {
                v.push((resolve_key(&e.key, heap), heap[e.val].clone()));
            }
        }
        Filler::Map(rc, pairs) => {
            let mut v = rc.borrow_mut();
            for (k, val) in pairs {
                v.push((heap[k].clone(), heap[val].clone()));
            }
        }
        Filler::Set(rc, refs) => {
            let mut v = rc.borrow_mut();
            for i in refs {
                v.push(heap[i].clone());
            }
        }
        Filler::Error(rc, cause_ref, entries) => {
            let mut e = rc.borrow_mut();
            if let Some(ci) = cause_ref {
                e.cause = heap[ci].clone();
            }
            for entry in entries {
                e.extra
                    .push((resolve_key(&entry.key, heap), heap[entry.val].clone()));
            }
        }
    }
}

pub fn decode(data: &[u8]) -> Result<Value, String> {
    let mut r = Reader::new(data);
    if r.take(4)? != b"GRF1" {
        return Err("bad magic: not a Graft file".into());
    }
    let version = r.u8()?;
    if version != 1 {
        return Err(format!("unsupported version: {}", version));
    }
    let root = r.uvarint_usize()?;
    let count = r.uvarint_usize()?;

    let mut heap: Vec<Value> = Vec::with_capacity(count);
    let mut fillers: Vec<Filler> = Vec::new();
    for _ in 0..count {
        let (value, filler) = read_node(&mut r)?;
        heap.push(value);
        if let Some(f) = filler {
            fillers.push(f);
        }
    }
    for f in fillers {
        fill(f, &heap);
    }
    heap.get(root)
        .cloned()
        .ok_or_else(|| format!("root index {} out of range", root))
}
